sling/nlp/document/document-source.cc

```diff
@@ -96,6 +97,55 @@ class ZipDocumentSource : public DocumentSource {
   int current_;
 };
 
+// Iterator implementation for SLING recordio files.
+// Assumes that each encoded document is a separate record in the recordio file.
+class RecordIODocumentSource : public DocumentSource {
+ public:
+  RecordIODocumentSource(const string &file) {
+    reader_ = new RecordReader(file);
+    file_ = file;
+  }
+
+  ~RecordIODocumentSource() override {
+    if (reader_ != nullptr) reader_->Close();
+    delete reader_;
+  }
+
+  bool NextSerialized(string *name, string *contents) override {
+    if (reader_->Done()) return false;
+
+    Record record;
+    CHECK(reader_->Read(&record));
+
+    *name = record.key.str();
+    *contents = record.value.str();
+
+    return true;
+  }
+
+  Document *Next(Store *store, string *name) override {
+    if (reader_->Done()) return false;
+
+    Record record;
+    CHECK(reader_->Read(&record));
+    *name = record.key.str();
+
+    StringDecoder decoder(store, record.value.data(), record.value.size());
+    return new Document(decoder.Decode().AsFrame());
+  }
+
+  void Rewind() override {
+    if (reader_ != nullptr) {
+      delete reader_;
+      reader_ = new RecordReader(file_);
+    }
+  }
+
+ private:
+  RecordReader *reader_ = nullptr;
```

Comment:
You could make this statically allocated since you are initializing it in the constructor.

Reply:
I tried that earlier, but the problem is in DocumentSource::Rewind(). In that method, simply calling RecordReader::Seek(0) doesn't work (it gives a corrupt header error during subsequent calls to ReadRecord()). Basically on rewinding I need to do most of the work that is currently done in the constructor of RecordReader, hence I am creating it again.